File: app/src/main/java/com/qimian233/ztool/utils/AppChooserDialog.java
package com.qimian233.ztool.utils;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Looper;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.qimian233.ztool.R;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AppChooserDialog {

    public interface AppSelectionCallback {
        void onSelected(List<AppInfo> selectedApps);
        void onCancel();
    }

    public static class AppInfo {
        private String packageName;
        private String appName;
        private Drawable appIcon;
        private boolean isSelected;

        public AppInfo(String packageName, String appName, Drawable appIcon) {
            this.packageName = packageName;
            this.appName = appName;
            this.appIcon = appIcon;
            this.isSelected = false;
        }

        public String getPackageName() { return packageName; }
        public String getAppName() { return appName; }
        public Drawable getAppIcon() { return appIcon; }
        public boolean isSelected() { return isSelected; }
        public void setSelected(boolean selected) { isSelected = selected; }
    }

    private static class AppAdapter extends RecyclerView.Adapter<AppAdapter.AppViewHolder> {
        private List<AppInfo> appList;
        private List<AppInfo> filteredList;
        private Set<String> selectedPackages;

        public AppAdapter(List<AppInfo> appList) {
            this.appList = appList;
            this.filteredList = new ArrayList<>(appList);
            this.selectedPackages = new HashSet<>();
            // 初始化 selectedPackages 基于 appList 中的选中状态
            for (AppInfo app : appList) {
                if (app.isSelected()) {
                    selectedPackages.add(app.getPackageName());
                }
            }
        }

        @NonNull
        @Override
        public AppViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_app_choice, parent, false);
            return new AppViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull AppViewHolder holder, int position) {
            AppInfo appInfo = filteredList.get(position);
            holder.bind(appInfo);
        }

        @Override
        public int getItemCount() {
            return filteredList.size();
        }

        public void filter(String query) {
            filteredList.clear();
            if (query == null || query.isEmpty()) {
                filteredList.addAll(appList);
            } else {
                String lowerCaseQuery = query.toLowerCase();
                for (AppInfo app : appList) {
                    if (app.getAppName().toLowerCase().contains(lowerCaseQuery) ||
                            app.getPackageName().toLowerCase().contains(lowerCaseQuery)) {
                        filteredList.add(app);
                    }
                }
            }
            notifyDataSetChanged();
        }

        public List<AppInfo> getSelectedApps() {
            List<AppInfo> selected = new ArrayList<>();
            for (AppInfo app : appList) {
                if (app.isSelected()) {
                    selected.add(app);
                }
            }
            return selected;
        }

        class AppViewHolder extends RecyclerView.ViewHolder {
            private ImageView appIcon;
            private TextView appName;
            private TextView packageName;
            private CheckBox checkBox;

            public AppViewHolder(@NonNull View itemView) {
                super(itemView);
                appIcon = itemView.findViewById(R.id.app_icon);
                appName = itemView.findViewById(R.id.app_name);
                packageName = itemView.findViewById(R.id.package_name);
                checkBox = itemView.findViewById(R.id.checkbox);
            }

            public void bind(AppInfo appInfo) {
                appIcon.setImageDrawable(appInfo.getAppIcon());
                appName.setText(appInfo.getAppName());
                packageName.setText(appInfo.getPackageName());
                checkBox.setChecked(appInfo.isSelected());

                itemView.setOnClickListener(v -> {
                    boolean newState = !appInfo.isSelected();
                    appInfo.setSelected(newState);
                    checkBox.setChecked(newState);

                    if (newState) {
                        selectedPackages.add(appInfo.getPackageName());
                    } else {
                        selectedPackages.remove(appInfo.getPackageName());
                    }
                });

                checkBox.setOnClickListener(v -> {
                    boolean newState = checkBox.isChecked();
                    appInfo.setSelected(newState);

                    if (newState) {
                        selectedPackages.add(appInfo.getPackageName());
                    } else {
                        selectedPackages.remove(appInfo.getPackageName());
                    }
                });
            }
        }
    }

    // 原有方法：不传入已勾选包名，保持兼容
    public static void show(Context context,
                            List<String> packageNames,
                            AppSelectionCallback callback) {
        show(context, packageNames, null, null, callback);
    }

    // 原有方法：传入标题，不传入已勾选包名
    public static void show(Context context,
                            List<String> packageNames,
                            String title,
                            AppSelectionCallback callback) {
        show(context, packageNames, null, title, callback);
    }

    // 新方法：传入已勾选包名列表
    public static void show(Context context,
                            List<String> packageNames,
                            List<String> selectedPackageNames,
                            AppSelectionCallback callback) {
        show(context, packageNames, selectedPackageNames, null, callback);
    }

    // 新方法：传入已勾选包名列表和标题
    public static void show(Context context,
                            List<String> packageNames,
                            List<String> selectedPackageNames,
                            String title,
                            AppSelectionCallback callback) {

        // 创建加载中对话框
        AlertDialog loadingDialog = new MaterialAlertDialogBuilder(context)
                .setTitle("加载中...")
                .setMessage("正在获取应用信息")
                .setCancelable(false)
                .create();
        loadingDialog.show();

        ExecutorService executor = Executors.newSingleThreadExecutor();
        Handler handler = new Handler(Looper.getMainLooper());

        executor.execute(() -> {
            List<AppInfo> appInfoList = new ArrayList<>();
            PackageManager pm = context.getPackageManager();

            for (String packageName : packageNames) {
                try {
                    ApplicationInfo appInfo = pm.getApplicationInfo(packageName, 0);
                    String appName = pm.getApplicationLabel(appInfo).toString();
                    Drawable appIcon = pm.getApplicationIcon(appInfo);

                    AppInfo app = new AppInfo(packageName, appName, appIcon);
                    // 根据 selectedPackageNames 设置选中状态
                    if (selectedPackageNames != null && selectedPackageNames.contains(packageName)) {
                        app.setSelected(true);
                    }
                    appInfoList.add(app);
                } catch (PackageManager.NameNotFoundException e) {
                    // 包名不存在，跳过
                }
            }

            handler.post(() -> {
                loadingDialog.dismiss();
                showAppSelectionDialog(context, appInfoList, title, callback);
            });
        });
    }

    private static void showAppSelectionDialog(Context context,
                                               List<AppInfo> appInfoList,
                                               String title,
                                               AppSelectionCallback callback) {

        View dialogView = LayoutInflater.from(context).inflate(R.layout.dialog_app_chooser, null);

        RecyclerView recyclerView = dialogView.findViewById(R.id.recycler_view);
        EditText searchEditText = dialogView.findViewById(R.id.search_edit_text);

        // 动态设置 RecyclerView 高度
        DisplayMetrics displayMetrics = context.getResources().getDisplayMetrics();
        int screenHeight = displayMetrics.heightPixels;
        int dialogMaxHeight = (int) (screenHeight * 0.6); // 屏幕高度的60%

        ViewGroup.LayoutParams params = recyclerView.getLayoutParams();
        params.height = dialogMaxHeight;
        recyclerView.setLayoutParams(params);

        AppAdapter adapter = new AppAdapter(appInfoList);
        recyclerView.setLayoutManager(new LinearLayoutManager(context));
        recyclerView.setAdapter(adapter);

        // 其余代码保持不变...
        searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                adapter.filter(s.toString());
            }

            @Override
            public void afterTextChanged(Editable s) {}
        });

        MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(context)
                .setView(dialogView)
                .setPositiveButton(context.getString(R.string.restart_yes), (dialog, which) -> {
                    if (callback != null) {
                        callback.onSelected(adapter.getSelectedApps());
                    }
                })
                .setNegativeButton(context.getString(R.string.restart_no), (dialog, which) -> {
                    if (callback != null) {
                        callback.onCancel();
                    }
                })
                .setOnCancelListener(dialog -> {
                    if (callback != null) {
                        callback.onCancel();
                    }
                });

        if (title != null) {
            builder.setTitle(title);
        }

        AlertDialog dialog = builder.create();

        // 设置对话框窗口属性，确保在横屏时也能正确显示
        Window window = dialog.getWindow();
        if (window != null) {
            window.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        }

        dialog.show();
    }
}

File: app/src/main/java/com/qimian233/ztool/utils/AppChooserDialog.kt
package com.qimian233.ztool.utils

import android.content.Context
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.graphics.drawable.Drawable
import android.os.Handler
import android.os.Looper
import android.app.Dialog
import android.view.ViewGroup
import android.widget.ImageView
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.viewinterop.AndroidView
import com.qimian233.ztool.R
import com.qimian233.ztool.ui.components.showPlatformComposeDialog
import java.util.concurrent.Executors

import com.qimian233.ztool.ui.components.ZToolDialogSurface

object AppChooserDialog {
    interface AppSelectionCallback {
        fun onSelected(selectedApps: List<AppInfo>)
        fun onCancel()
    }

    class AppInfo(
        val packageName: String,
        val appName: String,
        val appIcon: Drawable
    ) {
        var isSelected: Boolean = false
    }

    @JvmStatic
    fun show(
        context: Context,
        packageNames: List<String>,
        callback: AppSelectionCallback
    ) {
        show(context, packageNames, null, null, callback)
    }

    @JvmStatic
    fun show(
        context: Context,
        packageNames: List<String>,
        title: String?,
        callback: AppSelectionCallback
    ) {
        show(context, packageNames, null, title, callback)
    }

    @JvmStatic
    fun show(
        context: Context,
        packageNames: List<String>,
        selectedPackageNames: List<String>?,
        callback: AppSelectionCallback
    ) {
        show(context, packageNames, selectedPackageNames, null, callback)
    }

    @JvmStatic
    fun show(
        context: Context,
        packageNames: List<String>,
        selectedPackageNames: List<String>?,
        title: String?,
        callback: AppSelectionCallback?
    ) {
        val loadingDialog = showComposeDialog(context, cancelable = false) { _ ->
            LoadingContent(message = stringResource(R.string.loadingUserAPP))
        }

        val executor = Executors.newSingleThreadExecutor()
        val handler = Handler(Looper.getMainLooper())

        executor.execute {
            val appInfoList = loadApps(context, packageNames, selectedPackageNames)
            handler.post {
                loadingDialog.dismiss()
                executor.shutdown()
                showAppSelectionDialog(context, appInfoList, title, callback)
            }
        }
    }

    private fun loadApps(
        context: Context,
        packageNames: List<String>,
        selectedPackageNames: List<String>?
    ): List<AppInfo> {
        val packageManager = context.packageManager
        val selectedSet = selectedPackageNames?.toSet().orEmpty()
        val selectedApps = mutableListOf<AppInfo>()
        val unselectedApps = mutableListOf<AppInfo>()

        packageNames.forEach { packageName ->
            val app = packageManager.loadAppInfo(packageName) ?: return@forEach
            app.isSelected = packageName in selectedSet
            if (app.isSelected) {
                selectedApps += app
            } else {
                unselectedApps += app
            }
        }

        return if (selectedPackageNames == null) unselectedApps else selectedApps + unselectedApps
    }

    private fun PackageManager.loadAppInfo(packageName: String): AppInfo? {
        return try {
            val applicationInfo: ApplicationInfo = getApplicationInfo(packageName, 0)
            AppInfo(
                packageName = packageName,
                appName = getApplicationLabel(applicationInfo).toString(),
                appIcon = getApplicationIcon(applicationInfo)
            )
        } catch (_: PackageManager.NameNotFoundException) {
            null
        }
    }

    private fun showAppSelectionDialog(
        context: Context,
        appInfoList: List<AppInfo>,
        title: String?,
        callback: AppSelectionCallback?
    ) {
        lateinit var dialog: Dialog
        dialog = showComposeDialog(context, cancelable = true) {
            AppChooserContent(
                title = title,
                apps = appInfoList,
                onConfirm = { selectedPackages ->
                    appInfoList.forEach { app ->
                        app.isSelected = app.packageName in selectedPackages
                    }
                    callback?.onSelected(appInfoList.filter { it.isSelected })
                    dialog.dismiss()
                },
                onCancel = {
                    callback?.onCancel()
                    dialog.dismiss()
                }
            )
        }
        dialog.setOnCancelListener {
            callback?.onCancel()
        }
    }

    private fun showComposeDialog(
        context: Context,
        cancelable: Boolean,
        content: @Composable (Dialog) -> Unit
    ): Dialog {
        return showPlatformComposeDialog(
            context = context,
            cancelable = cancelable,
            width = ViewGroup.LayoutParams.MATCH_PARENT,
            height = ViewGroup.LayoutParams.WRAP_CONTENT,
            content = content
        )
    }
}

@Composable
private fun LoadingContent(message: String) {
    ZToolDialogSurface {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(24.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            CircularProgressIndicator(modifier = Modifier.size(28.dp))
            Spacer(modifier = Modifier.width(16.dp))
            Text(
                text = message,
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurface
            )
        }
    }
}

@Composable
private fun AppChooserContent(
    title: String?,
    apps: List<AppChooserDialog.AppInfo>,
    onConfirm: (Set<String>) -> Unit,
    onCancel: () -> Unit
) {
    var query by remember { mutableStateOf("") }
    val selectedPackages = remember {
        mutableStateListOf<String>().apply {
            addAll(apps.filter { it.isSelected }.map { it.packageName })
        }
    }
    val filteredApps = remember(query, apps, selectedPackages.size) {
        val normalizedQuery = query.trim().lowercase()
        if (normalizedQuery.isEmpty()) {
            apps
        } else {
            apps.filter { app ->
                app.appName.lowercase().contains(normalizedQuery) ||
                    app.packageName.lowercase().contains(normalizedQuery)
            }
        }
    }

    ZToolDialogSurface {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = 20.dp, vertical = 16.dp)
        ) {
            if (!title.isNullOrBlank()) {
                Text(
                    text = title,
                    style = MaterialTheme.typography.titleLarge,
                    color = MaterialTheme.colorScheme.onSurface,
                    fontWeight = FontWeight.SemiBold
                )
                Spacer(modifier = Modifier.height(12.dp))
            }

            OutlinedTextField(
                value = query,
                onValueChange = { query = it },
                modifier = Modifier.fillMaxWidth(),
                singleLine = true,
                label = { Text(stringResource(R.string.SearchHint)) }
            )

            Text(
                text = selectedPackages.size.toString(),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 8.dp)
            )

            LazyColumn(
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(max = 420.dp)
                    .padding(top = 8.dp)
            ) {
                items(
                    items = filteredApps,
                    key = { it.packageName }
                ) { app ->
                    val selected = app.packageName in selectedPackages
                    AppChooserRow(
                        app = app,
                        selected = selected,
                        onSelectedChange = { checked ->
                            if (checked) {
                                if (app.packageName !in selectedPackages) {
                                    selectedPackages += app.packageName
                                }
                            } else {
                                selectedPackages -= app.packageName
                            }
                        }
                    )
                }
            }

            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 12.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Spacer(modifier = Modifier.weight(1f))
                TextButton(onClick = onCancel) {
                    Text(stringResource(R.string.restart_no))
                }
                TextButton(onClick = { onConfirm(selectedPackages.toSet()) }) {
                    Text(stringResource(R.string.restart_yes))
                }
            }
        }
    }
}

@Composable
private fun AppChooserRow(
    app: AppChooserDialog.AppInfo,
    selected: Boolean,
    onSelectedChange: (Boolean) -> Unit
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onSelectedChange(!selected) }
            .padding(vertical = 10.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        AndroidView(
            modifier = Modifier.size(44.dp),
            factory = { context ->
                ImageView(context).apply {
                    scaleType = ImageView.ScaleType.FIT_CENTER
                    contentDescription = context.getString(R.string.app_icon)
                }
            },
            update = { imageView ->
                imageView.setImageDrawable(app.appIcon)
            }
        )
        Spacer(modifier = Modifier.width(14.dp))
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text = app.appName,
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.onSurface,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
            Text(
                text = app.packageName,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis
            )
        }
        Checkbox(
            checked = selected,
            onCheckedChange = onSelectedChange
        )
    }
}
